catfs/fs.go

```diff
@@ -955,35 +955,36 @@ func (fs *FS) Stage(path string, r io.Reader) error {
 		fs.cfg.String("compress.default_algo"),
 	)
 
+	// We might not be able to read the header for some reason.
+	// It's not critical currently, so fall back to the normal reader.
+	baseReader := r
+
 	if err != nil {
 		log.WithError(err).Warnf("failed to read default compression, using default")
 		compressAlgo = compress.AlgoNone
-	}
-
-	// Keep the header of the file in memory, so we can do some guessing
-	// of e.g. the compression algorithm we should use.
-	headerReader := util.NewHeaderReader(r, 4096)
-	headerBuf, err := headerReader.Peek()
-	if err != nil {
-		log.WithError(err).Warnf("failed to peek stream header")
 	} else {
-		compressAlgo, err = compress.GuessAlgorithm(
-			path,
-			headerBuf,
-		)
-
+		// Keep the header of the file in memory, so we can do some guessing
+		// of e.g. the compression algorithm we should use.
+		headerReader := util.NewHeaderReader(r, 2048)
+		headerBuf, err := headerReader.Peek()
 		if err != nil {
-			log.WithError(err).Warnf("failed to guess suitable zip algo for %s", path)
```

Comment:
Would failed attempt of `headerReader.Peek()` move the cursor position in `r`? Shall we reset it back to zero?

Reply:
Yes, it does move the cursor position, but headerReader preprends the buffer it read back to the stream.